dmutils/email/dm_notify.py

```diff
@@ -0,0 +1,123 @@
+from notifications_python_client.errors import HTTPError
+
+from dmutils.email.exceptions import DMNotifyError
+from dmutils.email.helpers import hash_string
+from dmutils.formats import stringify_dict
+
+
+class DMNotifyClient(object):
+    """Digital Marketplace wrapper around the Notify python client."""
+
+    _client_class = NotificationsAPIClient
+    _sent_references_cache = None
+
+    def __init__(
+            self,
+            govuk_notify_api_key,
+            govuk_notify_base_url='https://api.notifications.service.gov.uk',
+            logger=None
+    ):
+        """Set up logging and mail client."""
+        self.logger = logger or current_app.logger
+        self.client = self._client_class(govuk_notify_api_key, govuk_notify_base_url)
+
+    def get_all_notifications(self, **kwargs):
+        """Wrapper for notifications_python_client.notifications.NotificationsAPIClient::get_all_notifications"""
+        return self.client.get_all_notifications(**kwargs)['notifications']
+
+    def get_delivered_notifications(self, **kwargs):
+        """Wrapper for notifications_python_client.notifications.NotificationsAPIClient::get_all_notifications"""
+        return self.get_all_notifications(status='delivered', **kwargs)
+
+    def get_delivered_references(self, invalidate_cache=False):
+        """Get the references of all notifications that have already been delivered."""
+        if invalidate_cache:
+            self._sent_references_cache = None
+        if self._sent_references_cache is None:
+            self._sent_references_cache = [i['reference'] for i in self.get_delivered_notifications()]
+        return self._sent_references_cache
+
+    def _update_cache(self, reference):
+        """If the cache has been instantiated then cache the new reference."""
+        if self._sent_references_cache is not None:
+            self._sent_references_cache.append(reference)
+
+    def has_been_sent(self, reference):
+        """Checks for a matching reference in our list of delivered references."""
+        return reference in self.get_delivered_references()
+
+    @staticmethod
+    def get_reference(email_address, template_id, personalisation=None, **kwargs):
+        """
+        Method to return the standard reference given the variables the email is sent with.
+
+        :param email_address: Emails recipient
+        :param template_id: Emails template ID on Notify
+        :param personalisation: Template parameters
+        :param kwargs: Extra data passed to the reference eg. {'notes': 'Manual resend'}
+        :return: Hashed string 'reference' to be passed to client.send_email_notification or self.send_email
+        """
+        details_dict = OrderedDict([
+            ('email_address', email_address),
+            ('template_id', template_id),
+            ('personalisation', ','.join(stringify_dict(personalisation))),
+        ])
+        details_string = '|'.join(stringify_dict(details_dict))
```

Comment:
Fair, we'll likely still need OrderedDict to ensure ordering for the tests and so that the ref is created the same but yeah.